vamb/reclustering.py

```diff
@@ -0,0 +1,319 @@
+        op = lambda f,_ : f
+    elif fname.endswith('.gz'):
+        import gzip
+        op = gzip.open
+    elif fname.endswith('.bz2'):
+        import bz2
+        op = bz2.open
+    elif fname.endswith('.xz'):
+        import lzma
+        op = lzma.open
+    else:
+        op = open
+    with op(fname, 'rt') as f:
+        for line in f:
+            if line[0] == '>':
+                if header is not None:
+                    yield header,''.join(chunks)
+                line = line[1:].strip()
+                if not line:
+                    header = ''
+                elif full_header:
+                    header = line.strip()
+                else:
+                    header = line.split()[0]
+                chunks = []
+            else:
+                chunks.append(line.strip())
+        if header is not None:
+            yield header, ''.join(chunks)
+
+
+normalize_marker_trans__dict = {
+    'TIGR00388': 'TIGR00389',
+    'TIGR00471': 'TIGR00472',
+    'TIGR00408': 'TIGR00409',
+    'TIGR02386': 'TIGR02387',
+}
+
+
+def get_marker(hmmout, fasta_path=None, min_contig_len=None, multi_mode=False, orf_finder = None, contig_to_marker = False):
+    '''Parse HMM output file and return markers
+    '''
+    data = pd.read_table(hmmout, sep=r'\s+',  comment='#', header=None,
+                         usecols=(0,3,5,15,16), names=['orf', 'gene', 'qlen', 'qstart', 'qend'])
+    if not len(data):
+        return []
+    data['gene'] = data['gene'].map(lambda m: normalize_marker_trans__dict.get(m , m))
+    qlen = data[['gene','qlen']].drop_duplicates().set_index('gene')['qlen']
+
+    def contig_name(ell):
+        if orf_finder in ['prodigal', 'fast-naive']:
+            contig,_ = ell.rsplit( '_', 1)
+        else:
+            contig,_,_,_ = ell.rsplit( '_', 3)
+        return contig
+
+    data = data.query('(qend - qstart) / qlen > 0.4').copy()
+    data['contig'] = data['orf'].map(contig_name)
+    if min_contig_len is not None:
+        contig_len = {h:len(seq) for h,seq in fasta_iter(fasta_path)}
+        data = data[data['contig'].map(lambda c: contig_len[c] >= min_contig_len)]
+    data = data.drop_duplicates(['gene', 'contig'])
+
+    if contig_to_marker:
+        from collections import defaultdict
+        marker = data['gene'].values
+        contig = data['contig'].str.split('.').str[-1].values
+        sequence2markers = defaultdict(list)
+        for m, c in zip(marker, contig):
+            sequence2markers[c].append(m)
+        return sequence2markers
+    else:
+        def extract_seeds(vs, sel):
+            vs = vs.sort_values()
+            median = vs[len(vs) //2]
+
+            # the original version broke ties by picking the shortest query, so we
+            # replicate that here:
+            candidates = vs.index[vs == median]
+            c = qlen.loc[candidates].idxmin()
+            r = list(sel.query('gene == @c')['contig'])
+            r.sort()
+            return r
```

Comment:
Can this be factored out to its own top level function? Nested functions are not a good idea, IMO